### PR TITLE
Refactoring binary broadcasting

### PR DESCRIPTION
**Pull request by sxjscience**
Support ndim up to 7 for binary broadcasting operators + Accelerate reducing OPs by calling reduce_except_dim if possible. + Add `/bigobj` to CMakeList

### CODE REVIEW

**Comment by sxjscience** [comment]
@mli @tqchen

File: tests/python/unittest/test_operator.py [hunk 1]
```diff
@@ -733,24 +733,27 @@ def test_convolution_grouping():
...
+        r_axis_flags = np.ones(ndim)
+    l_shape = shape.copy()
+    r_shape = shape.copy()
+    l_shape[np.where(l_axis_flags == 0)] = 1
+    r_shape[np.where(r_axis_flags == 0)] = 1
+    return [np.random.random(l_shape), np.random.random(r_shape)]
 
 def _check_broadcast_op_forward(symbol, baseline):
-    for d in _gen_broadcast_data():
+    sample_num = 1000
```
Comment (piiswrong) [inline tests/python/unittest/test_operator.py]: how long does it take? Is 1000 necessary?

File: CMakeLists.txt [hunk 2]
```diff
@@ -29,7 +29,7 @@ if(MSVC)
   add_definitions(-D_CRT_SECURE_NO_WARNINGS)
   add_definitions(-DMXNET_EXPORTS)
   set(CMAKE_C_FLAGS "/MP")
-  set(CMAKE_CXX_FLAGS "${CMAKE_C_FLAGS}")
+  set(CMAKE_CXX_FLAGS "${CMAKE_C_FLAGS} /bigobj")
```
Comment (piiswrong) [inline CMakeLists.txt]: what's this?

File: tests/python/unittest/test_operator.py (hunk 1 above)
Comment (sxjscience) [inline tests/python/unittest/test_operator.py]: On my machine it's several seconds. I'd like to generate random tensors up to ndim=7 here. The sample number can be reduced.

Comment (sxjscience) [inline tests/python/unittest/test_operator.py]: @piiswrong I've reduced it to 200.

**Comment by tqchen** [comment]
@sxjscience  is there any way to get rid of the RANGE_SWITCH so we can directly make it to some cases like broadcast_mid? This will make the code more general and smaller

**Comment by sxjscience** [comment]
@tqchen, I may have to first get_shape_with a large NDIM like 7 and perform the reducing/broadcasting afterwards. RANGE_SWITCH can be avoid in this way.

**Comment by tqchen** [comment]
previously the reduce_mid and broadcast makes all solve the problem under three dimension array. I guess it is possible to reduce all problem to three dimension array? Collapsing other dimensions before and after the target axis

**Comment by sxjscience** [comment]
@tqchen The major problem here is the handling of non contiguous case. E.g, broadcasting (5,1,3,1,2) with (5,4,3,4,2). We need to broadcast over multiple axes that are separated.

**Comment by tqchen** [comment]
I see, so can we do the mid solution for one axis, and the range switch for the non continguous cases? So broadcast on one axis works for any dimensions

**Comment by sxjscience** [comment]
Yes, this is a good solution. Also, for the non-contiguous case, I can remove the RANGE_SWITCH by get shape with a fixed large NDIM (by padding ones to the shape vector). This seems to be acceptable and will not cause too many additional computation.

**Comment by tqchen** [comment]
sounds good

**Comment by tqchen** [comment]
let me know when you updated the code so we can merge it in

**Comment by sxjscience** [comment]
@tqchen I've changed the code. The data are read with ndim=3 if possible. For general case, I use ndim=7.

**Comment by tqchen** [comment]
please remove the rand switch for now since we discourage the use of it. Then i will merge it in

**Comment by sxjscience** [comment]
@tqchen I've removed MXNET_RANGE_SWITCH and rebased.

**Pull request merged**